stl/inc/semaphore

```diff
@@ -0,0 +1,304 @@
+        _Counter.notify_one();
+    }
+
+    void acquire() noexcept /* strengthened */ {
+        for (;;) {
+            // TRANSITION, GH-1133: should be memory_order_acquire
+            int8_t _Prev = _Counter.exchange(0, memory_order_seq_cst);
+            if (_Prev == 1) {
+                break;
+            }
+            _STL_VERIFY(_Prev == 0, "Invariant: semaphore counter is non-negative and not exceeds max(), "
+                                    "possibly caused by preconditions violation (N4861 [thread.sema.cnt]/8)");
+            _Counter.wait(0, memory_order_relaxed);
+        }
+    }
+
+    _NODISCARD bool try_acquire() noexcept {
+        // TRANSITION, GH-1133: should be memory_order_acquire
+        int8_t _Prev = _Counter.exchange(0, memory_order_seq_cst);
+        _STL_VERIFY((_Prev & ~1) == 0, "Invariant: semaphore counter is non-negative and not exceeds max(), "
+                                       "possibly caused by preconditions violation (N4861 [thread.sema.cnt]/8)");
+        return reinterpret_cast<const bool&>(_Prev);
+    }
+
+    template <class _Rep, class _Period>
+    _NODISCARD bool try_acquire_for(const chrono::duration<_Rep, _Period>& _Rel_time) {
+        auto _Deadline = _Semaphore_deadline(_Rel_time);
+        for (;;) {
+            // TRANSITION, GH-1133: should be memory_order_acquire
+            int8_t _Prev = _Counter.exchange(0, memory_order_seq_cst);
+            if (_Prev == 1) {
+                return true;
+            }
+            _STL_VERIFY(_Prev == 0, "Invariant: semaphore counter is non-negative and not exceeds max(), "
+                                    "possibly caused by preconditions violation (N4861 [thread.sema.cnt]/8)");
+            const auto _Remaining_timeout = __std_atomic_wait_get_remaining_timeout(_Deadline);
+            if (_Remaining_timeout == 0) {
+                return false;
+            }
+            __std_atomic_wait_direct(&_Counter, &_Prev, sizeof(_Prev), _Remaining_timeout);
+        }
+    }
+
+    template <class _Clock, class _Duration>
+    _NODISCARD bool try_acquire_until(const chrono::time_point<_Clock, _Duration>& _Abs_time) {
+        for (;;) {
+            // TRANSITION, GH-1133: should be memory_order_acquire
+            int8_t _Prev = _Counter.exchange(0, memory_order_seq_cst);
+            if (_Prev == 1) {
+                return true;
+            }
+            _STL_VERIFY(_Prev == 0, "Invariant: semaphore counter is non-negative and not exceeds max(), "
+                                    "possibly caused by preconditions violation (N4861 [thread.sema.cnt]/8)");
+
+            const unsigned long _Remaining_timeout = _Semaphore_remaining_timeout(_Abs_time);
+            if (_Remaining_timeout == 0) {
+                return false;
+            }
+
+            __std_atomic_wait_direct(&_Counter, &_Prev, sizeof(_Prev), _Remaining_timeout);
+        }
+    }
+
+private:
+    atomic<int8_t> _Counter;

```

Comment:
I prefer `unsigned char` instead of `int8_t`; I think `u?int8_t` just add confusion because they aren't integer types, they're character types.